File: HexChat/banhelper.py
import hexchat

__module_name__ = "Banhelper"
__module_author__ = "TingPing"
__module_version__ = "5"
__module_description__ = "Simplifies banning and quieting"

wasop = False

def is_op():
	for user in hexchat.get_list('users'):
		if user.nick == hexchat.get_info('nick'):
			if user.prefix == '@':
				return True
			else:
				return False

def do_op(deop=False):
	global wasop
	chan = hexchat.get_info('channel')

	if not deop:
		if not is_op():
			wasop = False
			hexchat.command('cs op {}'.format(chan))
		else:
			wasop = True
	else:
		if not wasop:
			hexchat.command('cs deop {}'.format(chan))

def do_command(cmd):
	if is_op():
		hexchat.command(cmd)
		do_op(deop=True)
		return False

	return True
		
def get_mask(nick):
	invalid_chars = ('*', '?', '!', '@', '$')
	if any(char in nick for char in invalid_chars):
		return nick # It's already a mask.

	for user in hexchat.get_list('users'):
		if hexchat.nickcmp(user.nick, nick) == 0:
			if user.account:
				return '$a:{}'.format(user.account)
			elif user.host:
				host = user.host.split('@')[1]
				return '*!*@{}'.format(host)
			else:
				hexchat.command('whois {}'.format(nick))
				print('BH: User info not found, enable irc_who_join or try again.')
				return None

def ban_cb(word, word_eol, userdata):
	if len(word) > 1:
		mask = get_mask(word[1])

		if mask:
			do_op()

			if word[0] == 'ban':
				command = 'mode +b {}'.format(mask)
			elif word[0] == 'kickban':
				nick = word[1]
				chan = hexchat.get_info('channel')
				message = word_eol[2] if len(word_eol) > 2 else ""

				command = 'mode +b {}\r\nKICK {} {} :{}'.format(mask, chan, nick, message)
			elif word[0] == 'quiet':
				command = 'mode +q {}'.format(mask)

			hexchat.hook_timer(100, do_command, command)

		return hexchat.EAT_HEXCHAT
	else:			
		return hexchat.EAT_NONE
		
def unload_cb(userdata):
	print(__module_name__ + ' version ' + __module_version__ + ' unloaded.')

hexchat.hook_command('kickban', ban_cb) 
hexchat.hook_command('ban', ban_cb)
hexchat.hook_command('quiet', ban_cb)
hexchat.hook_unload(unload_cb)
print(__module_name__ + ' version ' + __module_version__ + ' loaded.')

File: HexChat/banhelper.lua
-- SPDX-License-Identifier: MIT
hexchat.register('BanHelper', '1', 'Simpllifies banning and quieting')

local function is_op ()
	local nick = hexchat.get_info('nick')
	for user in hexchat.iterate('users') do
		if hexchat.nickcmp(user.nick, nick) == 0 then
			return user.prefix:find('^[@&]')
		end
	end

	print('BanHelper: Warning, could not find self in userlist!')
	return false
end

local function get_mask (nick)
	if nick:find('[%*?!@%$]') then
		return nick -- Already a mask
	end

	for user in hexchat.iterate('users') do
		if hexchat.nickcmp(user.nick, nick) == 0 then
			if user.account then
				return '$a:' .. user.account
			elseif user.host then
				return '*!*@' .. user.host:match('@(.+)$')
			else
				print('BanHelper: Warning, user info not found, try enabling irc_who_join')
				return nil
			end
		end
	end
end

for _, command in pairs({'kickban', 'ban', 'quiet'}) do
	hexchat.hook_command(command, function (word, word_eol)
		if not word[2] then
			return hexchat.EAT_NONE
		end

		local caller = word[1]:lower()
		local channel = hexchat.get_info('channel')
		local mask = get_mask(word[2])
		local command

		if not mask then
			return hexchat.EAT_ALL
		end

		if caller == 'ban' then
			command = 'MODE +b ' .. mask
		elseif caller == 'kickban' then
			command = string.format('MODE +b %s\r\nKICK %s %s :%s', mask,
			                        channel, word[2], word_eol[3] or '')
		elseif caller == 'quiet' then
			command = 'MODE +q ' .. mask
		end

		local wasop = is_op()
		if not wasop then
			hexchat.command('cs op ' .. channel)
		end
		hexchat.hook_timer(100, function ()
			if is_op() then
				hexchat.command(command)
				if not wasop then
					hexchat.command('cs deop ' .. channel)
				end
				return false
			end

			return true
		end)

		return hexchat.EAT_HEXCHAT
	end)
end
